RetailCoder.VBE/Refactorings/Rename/RenameRefactoring.cs

```diff
@@ -38,358 +40,496 @@ public void Refactor()
                 }
             }
-            else if (_model.Target.DeclarationType == DeclarationType.Parameter && _model.Target.ParentDeclaration.DeclarationType.HasFlag(DeclarationType.Property))
-            {
-                var getter = _model.Target.DeclarationType == DeclarationType.PropertyGet
-                    ? _model.Target
-                    : GetProperty(_model.Target.ParentDeclaration, DeclarationType.PropertyGet);
+            return false;
+        }
 
-                var letter = _model.Target.DeclarationType == DeclarationType.PropertyLet
-                    ? _model.Target
-                    : GetProperty(_model.Target.ParentDeclaration, DeclarationType.PropertyLet);
+        private void Rename()
+        {
+            if (_model == null || _model.Declarations == null || _model.Target == null) { return; }
 
-                var setter = _model.Target.DeclarationType == DeclarationType.PropertySet
-                    ? _model.Target
-                    : GetProperty(_model.Target.ParentDeclaration, DeclarationType.PropertySet);
+            if (UserCancelsRename()) { return; }
 
-                var properties = new[] {getter, letter, setter};
+            var handler = GetHandler(_model);
 
-                var parameters = _model.Declarations.Where(d =>
-                    d.DeclarationType == DeclarationType.Parameter &&
-                    properties.Contains(d.ParentDeclaration) &&
-                    d.IdentifierName == _model.Target.IdentifierName);
-
-                foreach (var param in parameters)
-                {
-                    RenameUsages(param);
-                    RenameDeclaration(param, _model.NewName);
-                }
+            try
+            {
+                handler.Rename();
             }
-            else
+            catch (Exception)
             {
-                RenameUsages(_model.Target);
+                _messageBox.Show(handler.ErrorMessage, RubberduckUI.RenameDialog_Caption);
             }
+        }
 
-            if (ModuleDeclarationTypes.Contains(_model.Target.DeclarationType))
+        //todo: Factory?
+        private static IRenameRefactoringHandler GetHandler(RenameModel model)
+        {
+            IRenameRefactoringHandler handler = null;
+            if (model.Target.DeclarationType.HasFlag(DeclarationType.Property))
+            {
+                handler = new RenamePropertyHandler(model);
+            }
+            else if (model.Target.DeclarationType == DeclarationType.Parameter
+                        && model.Target.ParentDeclaration.DeclarationType.HasFlag(DeclarationType.Property))
+            {
+                handler = new RenamePropertyParameterHandler(model);
+            }
+            else if (model.Target.DeclarationType.HasFlag(DeclarationType.Module))
+            {
+                handler = new RenameModuleHandler(model);
+            }
+            else if (model.Target.DeclarationType == DeclarationType.Project)
```

Comment:
Enum values are essentially numeric constants with a name. No declaration will have `DeclarationType == DeclarationType.Property`, but all properties will have `DeclarationType.HasFlag(DeclarationType.Property)`. Refer to the `DeclarationType` enum for what's contained in what.